torch/onnx/_internal/exporter.py

```diff
@@ -705,25 +706,44 @@ def __call__(
         Returns:
             The model output as computed by ONNX Runtime
         """
-        import onnxruntime  # type: ignore[import]
 
-        # model specified by the user has precedence, when specified
-        model_with_state_dict = model_with_state_dict or self._model_torch
+        # TODO: If ONNX used absolute paths on the initializers external data files,
+        # users could call ONNXProgram.save and use ONNXProgram.__call__ without the internal save below
+        with contextlib.ExitStack() as stack:
+            if self.fake_context:
+                tmpdir_path = stack.enter_context(tempfile.TemporaryDirectory())
+                warnings.warn(
+                    "Cannot run model directly from `ONNXProgram` because"
+                    " the model was exported using `enable_fake_mode`."
+                    " The model will be serialized to disk using a temporary folder ({tmpdir_path})"
+                    " to populate the model with initializers before being execution."
+                )
+                # TODO: Revisit the need of `model_with_state_dict` being a real model and not just its state
+                onnx_model = os.path.join(tmpdir_path, "model.onnx")
```

Comment:
Not sure if I understand correctly. Is the scenario that, the weights are already serialized to somewhere, so the onnx model itself is very small, and all weight tensor protos are pointing to external data through a relative path?

And a few more questions:
- Are the weights stored in the same temp folder, so that serializing model there resolves the core restriction of model and weights being in the same folder?
- Why do we still need the following `adapt_torch_inputs_to_onnx` step which seemingly appends weights to model inputs?

Reply:
In this scenario, the model has no initializers (because it came from fake mode), but the initializers WILL be serialized into external data on a temporary folder. The API assumes - for now - that `enable_fake_mode` captured a checkpoint file during export (at `ONNXProgram.fake_context._model_state_dicts`). Models and initializers are saved in the same temp folder - and deleted after `__call__` returns.



Two things:
A) When fake mode is not used, `adapt_torch_inputs_to_onnx` is needed to append initializer to model. 
B) This PR is only making easy the life of user to run fake models. It ignores the scenari in  which users do want to specify initializers as input

All APIs will probably will need revision to better accomodate A) and B)